### PR TITLE
[mle] update number parent requests in an attach cycle

### PR DESCRIPTION
**Pull request by abtink**
This commit updates the MLE attach process so that in the first attach
cycle device tries a total of six MLE Parent Requests, the first two
to routers only followed by four to routers and REEDs. For example,
the six Parent Request message will be used before device can decide
to act the leader. An MTD in the next attach attempt(if cannot find a
parent in first attempt cycle), will go to the model of two Parent
Requests (first to routers, then to routers/REEDs).

This change impacts the time it takes for a device to start as leader
(due to increased number of Parent Request and wait time). This commit
updates different test scripts to address the change in the wait
time.

- Related: openthread/ot-commissioner#228

------------

_Background_:
- This is related to [SPEC-1036](https://threadgroup.atlassian.net/browse/SPEC-1036).
- This change also addresses a specific situation which I noticed could happen with
  the existing implementation: When `mAttachMode` is `kSamePartition` we always
  send Parent Req to routers and REEDs even in `kAttachStateParentRequestRouter`
  but then during the check of received Parent Responses we may not accept a 
  parent with link quality lower than 3 since we think we are in the first attempt and 
  the Parent Req msg was to routers only.  The same situation cannot happen with 
  the implementation in this PR.

### CODE REVIEW

File: src/core/thread/mle.cpp [hunk 1]
```diff
@@ -1290,10 +1290,63 @@ void Mle::HandleAttachTimer(Timer &aTimer)
...
+Error Mle::DetermineParentRequestType(ParentRequestType &aType) const
+{
+    // This method determines the Parent Request type to use during an
+    // attach cycle based on `mAttachMode`, `mAttachCounter` and
+    // `mParentRequestCounter`. This method MUST be used while in
+    // `kAttachStateParentRequest` state.
+    //
+    // On success it returns `kErrorNone` and sets `aType`. It returns
+    // `kErrorNotFound` to indicate that device can now transition
+    // to`kAttachStateParentRequest`(has already sent required number
```
Comment (jwhui) [inline src/core/thread/mle.cpp:1302]: This method already requires `kAttachStateParentRequest`. What does it mean to transition to `kAttachStateParentRequest`?

```suggestion
    // to `kAttachStateParentRequest` (has already sent required number
```

File: tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py [hunk 2]
```diff
@@ -227,7 +227,7 @@ def test(self):
         self.simulator.go(config.ROUTER_STARTUP_DELAY)
         self.assertEqual(self.nodes[ROUTER23].get_state(), 'router')
 
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py:230]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/Cert_5_6_06_NetworkDataExpiration.py [hunk 3]
```diff
@@ -98,7 +98,7 @@ def test(self):
         self.simulator.set_lowpan_context(2, '2001:2:0:2::/64')
         self.simulator.set_lowpan_context(3, '2001:2:0:3::/64')
 
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_6_06_NetworkDataExpiration.py:101]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/Cert_5_6_06_NetworkDataExpiration.py [hunk 4]
```diff
@@ -118,7 +118,7 @@ def test(self):
 
         self.nodes[ROUTER].remove_prefix('2001:2:0:3::/64')
         self.nodes[ROUTER].register_netdata()
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_6_06_NetworkDataExpiration.py:121]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/Cert_5_6_06_NetworkDataExpiration.py [hunk 5]
```diff
@@ -137,7 +137,7 @@ def test(self):
                 self.assertTrue(self.nodes[LEADER].ping(addr))
 
         self.nodes[ROUTER].stop()
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_6_06_NetworkDataExpiration.py:140]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/Cert_5_6_09_NetworkDataForwarding.py [hunk 6]
```diff
@@ -106,7 +106,7 @@ def test(self):
         # Set lowpan context of sniffer
         self.simulator.set_lowpan_context(1, '2001:2:0:1::/64')
 
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_6_09_NetworkDataForwarding.py:109]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/Cert_5_7_01_CoapDiagCommands.py [hunk 7]
```diff
@@ -119,7 +119,7 @@ def test(self):
 
         for i in range(3, 7):
             self.nodes[i].start()
-            self.simulator.go(10)
+            self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_7_01_CoapDiagCommands.py:122]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/Cert_8_1_01_Commissioning.py [hunk 8]
```diff
@@ -78,15 +78,15 @@ class Cert_8_1_01_Commissioning(thread_cert.TestCase):
...
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
         self.assertEqual(self.nodes[COMMISSIONER].get_state(), 'leader')
         self.nodes[COMMISSIONER].commissioner_start()
         self.simulator.go(3)
         self.nodes[COMMISSIONER].commissioner_add_joiner(self.nodes[JOINER].get_eui64(), 'PSKD01')
 
         self.nodes[JOINER].interface_up()
         self.nodes[JOINER].joiner_start('PSKD01')
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_8_1_01_Commissioning.py:89]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/Cert_8_1_06_Commissioning.py [hunk 9]
```diff
@@ -86,7 +86,7 @@ def test(self):
 
         self.nodes[JOINER].interface_up()
         self.nodes[JOINER].joiner_start(PSKD, URL_2)
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_8_1_06_Commissioning.py:89]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/Cert_8_2_02_JoinerRouter.py [hunk 10]
```diff
@@ -74,7 +74,7 @@ def test(self):
 
         self.nodes[JOINER_ROUTER].interface_up()
         self.nodes[JOINER_ROUTER].joiner_start('PSKD01')
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_8_2_02_JoinerRouter.py:77]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/Cert_8_2_05_JoinerRouter.py [hunk 11]
```diff
@@ -119,7 +119,7 @@ def test(self):
 
         self.nodes[JOINER].interface_up()
         self.nodes[JOINER].joiner_start(PSKD, URL_2)
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_8_2_05_JoinerRouter.py:122]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/Cert_9_2_06_DatasetDissemination.py [hunk 12]
```diff
@@ -171,15 +171,15 @@ def test(self):
         steering_data_tlv = mesh_cop.SteeringData(bytes([0xff]))
         # Step 2
         self.nodes[COMMISSIONER].commissioner_mgmtset_with_tlvs([steering_data_tlv, commissioner_session_id_tlv])
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_9_2_06_DatasetDissemination.py:174]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/backbone/test_mlr_multicast_routing.py [hunk 13]
```diff
@@ -141,7 +141,7 @@ def test(self):
         self.simulator.go(5)
         self.assertEqual('router', self.nodes[PBBR].get_state())
         self.nodes[PBBR].enable_backbone_router()
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/backbone/test_mlr_multicast_routing.py:144]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/backbone/test_mlr_multicast_routing.py [hunk 14]
```diff
@@ -151,7 +151,7 @@ def test(self):
         self.simulator.go(5)
         self.assertEqual('router', self.nodes[SBBR].get_state())
         self.nodes[SBBR].enable_backbone_router()
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/backbone/test_mlr_multicast_routing.py:154]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/border_router/MATN/MATN_05_ReregistrationToSameMulticastGroup.py [hunk 15]
```diff
@@ -95,21 +95,21 @@ def test(self):
...
 
         br2.start()
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
         self.assertEqual('router', br2.get_state())
         self.assertFalse(br2.is_primary_backbone_router)
 
         host.start(start_radvd=False)
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/border_router/MATN/MATN_05_ReregistrationToSameMulticastGroup.py:112]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/border_router/MATN/MATN_12_HopLimitProcessing.py [hunk 16]
```diff
@@ -94,7 +94,7 @@ def test(self):
         self.assertEqual('router', router1.get_state())
 
         host.start(start_radvd=False)
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/border_router/MATN/MATN_12_HopLimitProcessing.py:97]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/border_router/test_advertising_proxy.py [hunk 17]
```diff
@@ -206,7 +206,7 @@ def test(self):
             client.srp_client_clear_service('my-service', '_ipps._tcp')
             client.srp_client_set_host_address(host_address)
             client.srp_client_add_service('my-service', '_ipps._tcp', service_port)
-            self.simulator.go(10)
+            self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/border_router/test_advertising_proxy.py:209]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/border_router/test_external_route.py [hunk 18]
```diff
@@ -142,7 +142,7 @@ def test(self):
         self.assertNotIn(ipaddress.IPv6Network(ROUTE1), br2_kernel_routes)
 
         br2.bash('ip link set eth0 down')
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/border_router/test_external_route.py:145]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/border_router/test_plat_udp_accessiblity.py [hunk 19]
```diff
@@ -114,7 +114,7 @@ def _testDhcp6ClientAfterReset(self, server, client, reset_device):
             self.simulator.go(3)
             server.register_netdata()
 
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/border_router/test_plat_udp_accessiblity.py:117]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/border_router/test_single_border_router.py [hunk 20]
```diff
@@ -162,7 +162,7 @@ def test(self):
         br.remove_prefix(ON_MESH_PREFIX2)
         br.register_netdata()
 
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/border_router/test_single_border_router.py:165]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/border_router/test_single_border_router.py [hunk 21]
```diff
@@ -199,7 +199,7 @@ def test(self):
 
         br.disable_br()
 
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/border_router/test_single_border_router.py:202]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/test_dataset_updater.py [hunk 22]
```diff
@@ -102,7 +102,7 @@ def test(self):
 
         # update initiated by ROUTER overridden by LEADER
         self.nodes[ROUTER].start_dataset_updater(channel=16)
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/test_dataset_updater.py:105]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/test_detach.py [hunk 23]
```diff
@@ -78,11 +78,11 @@ def test(self):
         child1 = self.nodes[CHILD_1]
 
         leader.start()
-        self.simulator.go(5)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
         self.assertEqual(leader.get_state(), 'leader')
 
         router1.start()
-        self.simulator.go(5)
+        self.simulator.go(config.ROUTER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/test_detach.py:85]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/test_on_mesh_prefix.py [hunk 24]
```diff
@@ -110,7 +110,7 @@ def test(self):
         # Set lowpan context of sniffer
         self.simulator.set_lowpan_context(3, '2002:2:0:3::/64')
 
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/test_on_mesh_prefix.py:113]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/v1_2_test_backbone_router_service.py [hunk 25]
```diff
@@ -214,7 +214,7 @@ def test(self):
         self.nodes[BBR_2].reset()
         self.nodes[LEADER_1_1].release_router_id(self.nodes[BBR_2].get_router_id())
         # Wait for the dissemination of Network Data without Backbone Router service
-        self.simulator.go(10)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
Comment (jwhui) [inline tests/scripts/thread-cert/v1_2_test_backbone_router_service.py:217]: Why `LEADER_STARTUP_DELAY` here?

File: tests/scripts/thread-cert/Cert_9_2_07_DelayTimer.py [hunk 26]
```diff
@@ -90,12 +90,12 @@ def test(self):
         self.simulator.go(5)
 
         self.nodes[ROUTER].start()
-        self.simulator.go(5)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
```
[the remark this answers is not in the document]
Comment (abtink) [reply tests/scripts/thread-cert/Cert_9_2_07_DelayTimer.py:93]: I think this one is correct :)

File: tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py (hunk 2 above)
Comment (abtink) [reply tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py:230]: Thanks Jonathan for going through it all and finding all cases. 😃

File: tests/scripts/thread-cert/test_mle_msg_key_seq_jump.py [hunk 27]
```diff
@@ -104,7 +104,7 @@ def test(self):
         self.assertEqual(reed.get_state(), 'child')
 
         router.start()
-        self.simulator.go(5)
+        self.simulator.go(config.ROUTER_STARTUP_DELAY)
```
[the remark this answers is not in the document]
Comment (abtink) [reply tests/scripts/thread-cert/test_mle_msg_key_seq_jump.py:107]: I think this one is ok as well... for router using `ROUTER_STARTUP_DELAY`

File: tests/scripts/thread-cert/test_detach.py [hunk 28]
```diff
@@ -145,10 +145,10 @@ def test(self):
         self.assertEqual(leader.get_state(), 'disabled')
 
         leader.start()
-        self.simulator.go(5)
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
         self.assertEqual(leader.get_state(), 'leader')
         router1.start()
-        self.simulator.go(5)
+        self.simulator.go(config.ROUTER_STARTUP_DELAY)
```
[the remark this answers is not in the document]
Comment (abtink) [reply tests/scripts/thread-cert/test_detach.py:151]: Same. For router using `ROUTER_STARTUP_DELAY`

File: tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py (hunk 2 above)
Comment (abtink) [reply tests/scripts/thread-cert/Cert_5_2_06_RouterDowngrade.py]: Thanks Jonathan. I tried to search in code to make sure all uses `LEADER_STARTUP_DELAY` are followed by state check for some node becoming leader. Used `git grep -A 1 "LEADER_STARTUP_DELAY"` then parsed the output to make sure the line after has the state check. Found a bunch of other instances of it being used incorrectly. Hopefully managed to find them all

**Review by jwhui — APPROVED** [review]

**Pull request merged**